pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return nil, err
+	}
+	// Find the besteffort pods we would need to evict if we already evicted the required guaranteed and burstable pods.
+	bestEffortToEvict, err := getPodsToPreemptByDistance(bestEffortPods, requirements.subtract(append(burstableToEvict, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	return append(append(bestEffortToEvict, burstableToEvict...), guarateedToEvict...), nil
+}
+
+// finds the pods that have pod requests >= admission requirements.
+// Chooses pods that minimize "distance" to the requirements.
+// If more than one pod exists that fulfills the remaining requirements,
+// it chooses the pod that has the "smaller resource request"
+// This method, by repeatedly choosing the pod that fulfills as much of the requirements as possible,
+// attempts to minimize the number of pods returned.
+func getPodsToPreemptByDistance(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	podsToEvict := []*v1.Pod{}
+	// evict pods by shortest distance from remaining requirements, updating requirements every round.
+	for len(requirements) > 0 {
+		if len(pods) == 0 {
+			return nil, fmt.Errorf("no set of running pods found to reclaim resources: %v", requirements.toString())
+		}
+		// all distances must be less than len(requirements), because the max distance for a single requirement is 1
+		bestDistance := float64(len(requirements) + 1)
+		bestPodIndex := 0
+		// Find the pod with the smallest distance from requirements
+		// Or, in the case of two equidistant pods, find the pod with "smaller" resource requests.
+		for i, pod := range pods {
+			dist := requirements.distance(pod)
+			if dist < bestDistance || (bestDistance == dist && smallerResourceRequest(pod, pods[bestPodIndex])) {
+				bestDistance = dist
+				bestPodIndex = i
+			}
+		}
+		// subtract the pod from requirements, and transfer the pod from input-pods to pods-to-evicted
+		requirements = requirements.subtract(pods[bestPodIndex])
+		podsToEvict = append(podsToEvict, pods[bestPodIndex])
+		pods[bestPodIndex] = pods[len(pods)-1]
+		pods = pods[:len(pods)-1]
+	}
+	return podsToEvict, nil
+}
+
+type admissionRequirement struct {
+	resourceName v1.ResourceName
+	quantity     int64
+}
+
+type admissionRequirementList []*admissionRequirement
+
+// distance of the pods requests from the admissionRequirements.
+// distance is measured by the fraction of the requirement satisfied by the pod,
+// so that each requirement is weighted equally, regardless of absolute magnitude.
+func (a admissionRequirementList) distance(pod *v1.Pod) float64 {
+	dist := float64(0)
+	for _, req := range a {
+		remainingRequest := float64(req.quantity - v1.GetResourceRequest(pod, req.resourceName))
+		if remainingRequest < 0 {
+			remainingRequest = 0
+		}
+		dist += math.Pow(remainingRequest/float64(req.quantity), 2)
```

Comment:
Vish asked the same question.  You seem to have two suggestions:
1. Normalize distances, by dividing by the total number.
2. Dont square distances.

1 is not necessary, since they are being compared against each other.  Dividing everything by 3, if we have 3 requirements, still gives the same comparison results.

2 Is up for debate.  The following is/was my reasoning:
Say we have 2 requirements: 1000m cpu, and 1000Mi memory.
Say we could pick one of two pods to evict next (we may evict others not in this group to fulfill remaining requirements afterwards)
The two options are:
1.  Request = 500m cpu, 500Mi memory
2. Request = 1000m cpu

My assumption, which may not be accurate, is that it would be better to evict pod (1), than pod (2).  This is because I believe we are more likely to find a pods that has a small amount of many different resources, than a large amount of a single resource.  It penalizes pods that dont fulfill a requirement at all, and encourages preempting pods that make some progress in a number of areas.

Another rationale, is that there may be very few pods that have a certain request.  Say only 2 pods have any cpu requests, and we would need to evict them both as part of preemption.  Since no other pods meet the cpu requirement, the two pods that do would be preferred for preemption.  This would be even more noticeable if critical pods used GPU (which they dont), since usually only a small number of pods would use it.  This prioritizes freeing resources that are "scarcer" on a node.